xapian-core/weight/bm25plusweight.cc

```diff
@@ -0,0 +1,157 @@
+/** @file bm25plusweight.cc
+ * @brief Xapian::BM25PlusWeight class - the BM25+ probabilistic formula
+ */
+/* Copyright (C) 2016  Vivek Pal
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <config.h>
+
+#include "xapian/weight.h"
+
+#include "debuglog.h"
+#include "omassert.h"
+#include "serialise-double.h"
+
+#include "xapian/error.h"
+
+#include <algorithm>
+#include <cmath>
+
+using namespace std;
+
+namespace Xapian {
+
+BM25PlusWeight *
+BM25PlusWeight::clone() const
+{
+    return new BM25PlusWeight(param_k1, param_k2, param_k3, param_b,
+			      param_min_normlen, param_delta);
+}
+
+void
+BM25PlusWeight::init(double factor)
+{
+    Xapian::doccount tf = get_termfreq();
+
+    double tw = 0;
+    // BM25+ formula gives termweight = (total_no_of_docs + 1) / tf
+    if (tf != 0) tw = (get_collection_size() + 1) / tf;
+
+    AssertRel(tw,>,0);
+
+    if (tw < 2) tw = tw * 0.5 + 1;
+    termweight = log(tw) * factor;
+    if (param_k3 != 0) {
+	double wqf_double = get_wqf();
+	termweight *= (param_k3 + 1) * wqf_double / (param_k3 + wqf_double);
+    }
+    termweight *= (param_k1 + 1);
+
+    LOGVALUE(WTCALC, termweight);
+
+}
```

Comment:
It would be better not to have a blank line at the end of the function body here.